init.lua

```diff
@@ -613,6 +507,108 @@ local function get_item_usages(item)
 	return usages
 end
 
+local function progressive_show_recipe(recipe, inv_items)
+	for _, item in pairs(recipe.items) do
+		local item_in_inv = false
+		if item:sub(1,6) == "group:" then
+			local groups = extract_groups(item)
+			for i = 1, #inv_items do
+				local item_groups = reg_items[inv_items[i]].groups
+				if item_has_groups(item_groups, groups) then
+					item_in_inv = true
+				end
+			end
+		else
+			for i = 1, #inv_items do
+				if inv_items[i] == item then
+					item_in_inv = true
+				end
+			end
+		end
+
+		if not item_in_inv then
+			return
+		end
+	end
+	return true
+end
+
+local function progressive_filter_recipes(recipes, player)
+	local invlist = player:get_inventory():get_list("main")
+	local inv_items = {}
+	for i = 1, #invlist do
+		local stack = invlist[i]
+		if not stack:is_empty() then
+			inv_items[#inv_items + 1] = stack:get_name()
+		end
+	end
+
+	if #inv_items == 0 then
+		return {}
+	end
+
+	local filtered = {}
+	for i = 1, #recipes do
+		local recipe = recipes[i]
+		if progressive_show_recipe(recipe, inv_items) then
+			filtered[#filtered + 1] = recipe
+		end
+	end
+
+	return filtered
+end
+
+local function get_progressive_items(player)
+	local items = {}
+	for i = 1, #init_items do
+		local item = init_items[i]
+		local recipes = progressive_filter_recipes(get_recipes(item), player)
+		if #recipes > 0 then
+			items[#items + 1] = item
+		end
+	end
+
+	return items
+end
+
+local function init_data(player, name)
+	local progressive_items = progressive_mode and get_progressive_items(player)
+	player_data[name] = {
+		filter = "",
+		pagenum = 1,
+		iX = sfinv_only and 8 or DEFAULT_SIZE,
+		items = progressive_items or init_items,
+		progressive_items = progressive_items or nil
```

Comment:
`or nil` not necessary. Also change the variable name, either this one or line 575.

Reply:
`or nil` is in case `progressive_items` is false (this will happen if `progressive_mode` is false). Moved the `or nil` above to make it clearer.

Reply:
OK I prefer `nil` over `false` in this case.

Reply:
`progressive_mode` might be false because `settings:get_bool` might return false is what I mean